tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -4340,6 +4340,116 @@ LogicalResult lowerWithScheduleSparseSegmentMeanOpCPU(
+           << "root_ops[0] is not a lmhlo_disc::WhereOp";
+  }
+  auto where = dyn_cast<lmhlo_disc::WhereOp>(root_ops[0]);
+  if (!where) {
+    return dominant_op->emitError()
+           << "can not cast root_ops[0] to lmhlo_disc::WhereOp";
+  }
+
+  auto loc = where.getLoc();
+  OpBuilder b(root_ops.back());
+
+  b.setInsertionPoint(where.getOperation());
+
+  Value input = where.getInput();
+  Value index = where.getIndex();
+  Value num_output_elements = where.getNumOutputElements();
+
+  auto input_type = input.getType().cast<MemRefType>();
+  auto input_rank = input_type.getRank();
+  auto input_elem_type = input_type.getElementType();
+  auto output_elem_type = index.getType().cast<MemRefType>().getElementType();
+
+  Value zero =
+      b.create<arith::ConstantOp>(loc, b.getIntegerAttr(b.getIndexType(), 0));
+  Value one =
+      b.create<arith::ConstantOp>(loc, b.getIntegerAttr(b.getIndexType(), 1));
+  Value one_for_acc = b.create<arith::ConstantOp>(
+      loc, b.getIntegerAttr(b.getIntegerType(64), 1));
+  SmallVector<Value, 1> num_output_load_index = {
+      b.create<arith::ConstantIndexOp>(loc, 0)};
+  // memset num_output_elements to 0
+  b.create<memref::StoreOp>(loc,
+                            b.create<arith::ConstantOp>(
+                                loc, b.getIntegerAttr(b.getIntegerType(64), 0)),
+                            num_output_elements, num_output_load_index);
+
+  // for (i = 0; i < num_element; ++i)
+  llvm::SmallVector<scf::ForOp, 4> for_ops;
+  llvm::SmallVector<Value, 4> multidim_index;
+  llvm::SmallVector<Value, 4> multidim_index_casted;
+  for (int i = 0; i < input_rank; ++i) {
+    Value dim = b.create<memref::DimOp>(loc, input, i);
+    auto for_op = b.create<scf::ForOp>(loc, /* lowerBound */ zero,
+                                       /* upperBound */ dim, /* step */ one);
+    for_ops.push_back(for_op);
+    for_op.getBody()->clear();
+    b.setInsertionPointToStart(for_op.getBody());
+    Value loop_index = for_op.getInductionVar();
+    multidim_index.push_back(loop_index);
+    multidim_index_casted.push_back(
+        b.create<arith::IndexCastOp>(loc, output_elem_type, loop_index));
+  }
+  Value input_elem = b.create<memref::LoadOp>(loc, input, multidim_index);
+  // if (input[i] != 0)
+  Value is_zero;
+  // NOTE: make sure cmp value is the same type as input
+  // otherwise the generated IR will seems ok, however will get segfault
+  // in llvm::CmpInst::CmpInst
+  if (input_elem_type.isa<IntegerType>()) {
+    Value zero_for_cmp =
+        b.create<arith::ConstantOp>(loc, b.getIntegerAttr(input_elem_type, 0));
+    is_zero = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ne, input_elem,
+                                      zero_for_cmp);
+  } else if (input_elem_type.isa<FloatType>()) {
+    Value zero_for_cmp =
+        b.create<arith::ConstantOp>(loc, b.getFloatAttr(input_elem_type, 0));
+    is_zero = b.create<arith::CmpFOp>(loc, arith::CmpFPredicate::ONE,
```

Comment:
what is `ONE` for...

Reply:
According to mlir docs, `ONE` means `ordered not equal`，which is used for float type comparsion, however I am not quite sure which of `UNE` or `ONE` is more suitable(I don't quite get the exact meaning of these 2), after some trials `ONE` works the way I hope for this comparsion.

Reply:
OK, got it ...  considering precision, floating type comparison maybe actually unstable. Just like writing normal C++ code, we just don't use `==` directly to compare two floating point numbers, instead, to check if they're close enough (like `np.isclose`).
So, does TF's WhereOp do something special on floating-point type?

Reply:
See: https://floating-point-gui.de/errors/comparison/